### ISSUE TITLE
Doc2Vec  strange results when using most_similar

### ISSUE THREAD

**Issue by anurags1976** [RaRe-Technologies/gensim#1560]
Hi, I have a corpus of 300-400 documents. Using Doc2Vec, model was created. I then try to find most_similar document in the corpus for a test file. Test file as such contain garbage text. Only few words are actual dictionary words. I still get >50% similarity against a file in corpus even though as such both have no similarity.
Need inputs on same.
I have used dm=0 with dbow_words=1 with vector dimension 59.

**Comment by merjildo** [comment]
Could you share the result of the count of each document ranks with respect to the training corpus?

**Comment by anurags1976** [comment]
Hi, Is this what you are looking for
MOST (274, 0.555317759513855): «1.txt»
MEDIAN (193, 0.1945732980966568): «2.txt»
LEAST (38, -0.22431580722332): «3.txt»

Top 5
0 (274, 0.555317759513855): «1.txt»
1 (265, 0.5122318267822266): «11.txt»
2 (279, 0.46970635652542114): «111.txt»
3 (280, 0.46798962354660034): «1111.txt»
4 (278, 0.465034544467926): «11111.txt»
let me know

**Comment by merjildo** [comment]
Yeah, but for every document in training corpus, you should count how many were found in the first position (pos '0'). I think it should be more than 90% of documents. It is necessary just to check if the model is able to find similar documents.

**Comment by anurags1976** [comment]
Hi,
Its coming as below
Counter({0: 167, 1: 64, 2: 32, 3: 9, 4: 6, 6: 4, 8: 4, 7: 3, 5: 2, 9: 2, 10: 1})
I feel the spread is because corpus has documents broadly of 3 types.

So would you suggest I should create 3 different models rather than combined model?

**Comment by anurags1976** [comment]
Just to share the scope of what I am trying to do. My use case is to identify whether a new document which I am trying to add to my content library (corpus) already exists or if there are similar documents in my content library (the corpus). So basically find duplicates/similar docs. So for that I was trying to find most_similar against whole of corpus. Hence created a single model.

**Comment by merjildo** [comment]
I don't think so, it looks like you are doing something wrong. Are you using gensim.models.doc2vec.TaggedDocument to associate a tag with each document of the training corpus?. I had a similar issue before use TaggedDocument.

**Comment by anurags1976** [comment]
Yes I am using TaggedDocument. This is how it looks

def read_corpus(files, tokens_only=False):
    for i, fname in enumerate(files):
        listFileNames.append(fname)
        with smart_open.smart_open(fname, encoding="iso-8859-1") as f:
            lines = f.readlines()
            out = ""
            for line in lines:
                out = out + line.strip().decode('iso-8859-1')
            if tokens_only:
                yield gensim.utils.simple_preprocess(out)
            else:
                # For training data, add tags
                yield gensim.models.doc2vec.TaggedDocument(gensim.utils.simple_preprocess(out), [i])


train_corpus = list(read_corpus(trainFiles))
test_corpus = list(read_corpus(testFiles, tokens_only=True))

model = gensim.models.doc2vec.Doc2Vec(size=65, min_count=2, dm=0, hs=1, iter=100, negative=0)
model.build_vocab(train_corpus)

model.train(train_corpus, total_examples=model.corpus_count)

**Comment by anurags1976** [comment]
Indentation went for a toss. My bad. read_corpus method is till the line where I am tagging the document

**Comment by merjildo** [comment]
In addition, I would suggest you check the way you are preprocessing your data. I got better results removing repeated words, removing stop words and using word stems, among others.

**Comment by gojomo** [comment]
Doc2Vec doesn't work very well on tiny datasets (like a few hundred documents). To some extent, you might be able to get some value by reducing the vector-size and increasing the iterations (as you've done)... but published work typically involves tens-of-thousands to millions of documents. 

Separately, the similarity scores returned should not be viewed as a "percentage similar" or other absolute measure of similarity. They're merely relative, within a certain model, given certain training parameters, compared to other results. The ordinal rank of top-N results is more significant, or magnitude of difference between alternate pairs. 

So getting similarity values > 0.5 for a 'junk' document isn't surprising or concerning. In particular, the model training will (very vaguely) be learning a position of all known documents across the full coordinate-space – and none of the coordinate space will be reserved for idiosyncratic docs unlike those in the training space. If you then throw a random/'junk' document at the model for vector-inference, it will do its best to make a vector for it, and there will be some 'nearest' docs, and here'd be no a priori threshold (like "0.5") which means it's "similar" or "dissimilar" by human standards. (Indeed, the same 2 documents A and B could have wildly different similarity-scores when trained into models with different vector-sizes, or training-parameters, or other co-trained peer documents.) There'll only be a score that's better or worse than other potential documents. (Perhaps, with enough examples of what docs should be considered similar or not, you could learn a threshold specific to a model & corpus.)

**Comment by anurags1976** [comment]
Thanks a lot for your inputs. This is really helpful. On your point that Doc2Vec doesn't work well on small datasets would you recommend something else?

**Comment by gojomo** [comment]
You can use a stronger model (trained on more documents) to model smaller sets of documents. Simple bag-of-words representations or bag-of-ngrams may work fine. There's a metric for document-to-document similarity called "word mover's distance" – available in gensim word-vector classes like `KeyedVectors` – that can re-use word-vectors from a larger training to compare other texts. (It's a bit expensive to calculate but with small document counts, no problem!)

**Comment by anurags1976** [comment]
I will definitely look into it. Thanks a lot for all your inputs.
